diff --git a/nv-attestation-sdk-cpp/src/rim.cpp b/nv-attestation-sdk-cpp/src/rim.cpp
--- a/nv-attestation-sdk-cpp/src/rim.cpp
+++ b/nv-attestation-sdk-cpp/src/rim.cpp
@@ -61,7 +61,7 @@
     //    TCG from CORIM
     auto doc = nv_unique_ptr<xmlDoc>(xmlReadDoc(reinterpret_cast<const xmlChar *>(rim_data.c_str()), NULL, NULL, XML_PARSE_PEDANTIC | XML_PARSE_NONET));
     if (!doc) {
-        xmlErrorPtr xml_error = xmlGetLastError();
+        const xmlError *xml_error = xmlGetLastError();
         std::string error_msg = "Failed to parse RIM data as TCG XML file. ";
         if (xml_error != nullptr) {
             error_msg += "XML error: " + std::string(xml_error->message);
